src/controller/proxy/controller.go

```diff
@@ -65,31 +65,56 @@ type Controller interface {
 	// art is the ArtifactInfo which includes the tag or digest of the manifest
 	ProxyManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (distribution.Manifest, error)
 	// HeadManifest send manifest head request to the remote server
-	HeadManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, string, error)
+	HeadManifest(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface) (bool, *distribution.Descriptor, error)
+	// EnsureTag ensure tag for digest
+	EnsureTag(ctx context.Context, art lib.ArtifactInfo, tagName string) error
 }
 type controller struct {
-	blobCtl     blob.Controller
-	artifactCtl artifact.Controller
-	local       localInterface
-	cache       cache.Cache
+	blobCtl         blob.Controller
+	artifactCtl     artifact.Controller
+	local           localInterface
+	cache           cache.Cache
+	handlerRegistry map[string]ManifestCacheHandler
 }
 
 // ControllerInstance -- Get the proxy controller instance
 func ControllerInstance() Controller {
 	// Lazy load the controller
 	// Because LocalHelper is not ready unless core startup completely
 	once.Do(func() {
+		l := newLocalHelper()
 		ctl = &controller{
-			blobCtl:     blob.Ctl,
-			artifactCtl: artifact.Ctl,
-			local:       newLocalHelper(),
-			cache:       cache.Default(),
+			blobCtl:         blob.Ctl,
+			artifactCtl:     artifact.Ctl,
+			local:           newLocalHelper(),
+			cache:           cache.Default(),
+			handlerRegistry: NewCacheHandlerRegistry(l),
 		}
 	})
 
 	return ctl
 }
 
+func (c *controller) EnsureTag(ctx context.Context, art lib.ArtifactInfo, tagName string) error {
+	// search the digest in cache and query with trimmed digest
+	var trimmedDigest string
+	err := c.cache.Fetch(TrimmedManifestlist+art.Digest, &trimmedDigest)
+	if err != cache.ErrNotFound {
```

Comment:
Seems the error handling logic isn't correct?
What happens if the error is `cache.ErrNotFound`?

Reply:
updated